File: ios/Plugin/CapacitorCalendarPlugin.swift
import Foundation
import EventKitUI
import EventKit
import Capacitor

@objc(CapacitorCalendarPlugin)
public class CapacitorCalendarPlugin: CAPPlugin {
    private let eventStore = EKEventStore()
    private lazy var calendar = CapacitorCalendar(bridge: self.bridge, eventStore: self.eventStore)
    private lazy var reminders = CapacitorReminders(bridge: self.bridge, eventStore: self.eventStore)

    // MARK: - Utils

    private func tryCall(_ call: CAPPluginCall, source: String, _ op: () throws -> Void) {
        do {
            try op()
        } catch {
            CapacitorCalendarError.reject(call, error: error, source: source)
        }
    }

    private func tryAsyncCall(_ call: CAPPluginCall, source: String, _ op: () async throws -> Void) async {
        do {
            try await op()
        } catch {
            CapacitorCalendarError.reject(call, error: error, source: source)
        }
    }

    private func getNonEmptyString(_ call: CAPPluginCall, _ param: String, source: String) -> String? {
        guard let value = call.getString(param),
              !value.isEmpty else {
            CapacitorCalendarError.reject(call, type: .missingKey, source: source, data: param)
            return nil
        }

        return value
    }

    // MARK: - Permissions

    private func callHasAccess(_ call: CAPPluginCall, entityType: EntityType, accessType: AccessType, source: String) throws -> Bool {
        let eventTypeName = String(describing: entityType)
        let accessTypeName = String(describing: accessType)
        let permission = try doPermissionCheck(for: "\(accessTypeName)\(eventTypeName.capitalized)")

        if permission == "granted" {
            return true
        } else {
            CapacitorCalendarError.rejectWithNoAccess(call, entityType: entityType, accessType: accessType, source: source)
            return false
        }
    }

    private func doPermissionCheck(for alias: String) throws -> String {
        let permissionsState: [String: String]

        switch alias {
        case "readCalendar":
            permissionsState = calendar.checkAllPermissions()

        case "writeCalendar":
            permissionsState = calendar.checkAllPermissions()

        case "readReminders":
            permissionsState = reminders.checkAllPermissions()

        case "writeReminders":
            permissionsState = reminders.checkAllPermissions()

        default:
            throw CapacitorCalendarError(.invalidKey, source: #function, data: alias)
        }

        guard let permissionResult = permissionsState[alias] else {
            throw CapacitorCalendarError(.internalError, source: #function, data: "Invalid permission state")
        }

        return permissionResult
    }

    @objc public func checkPermission(_ call: CAPPluginCall) {
        guard let alias = getNonEmptyString(call, "alias", source: #function) else {
            return
        }

        tryCall(call, source: #function) {
            try call.resolve(["result": self.doPermissionCheck(for: alias)])
        }
    }

    @objc public func checkAllPermissions(_ call: CAPPluginCall) {
        let calendarPermissionsState = calendar.checkAllPermissions()
        let remindersPermissionsState = reminders.checkAllPermissions()
        call.resolve(calendarPermissionsState.merging(remindersPermissionsState) { (_, new) in new })
    }

    @objc public func requestWriteOnlyCalendarAccess(_ call: CAPPluginCall) {
        Task {
            await tryAsyncCall(call, source: #function) {
                let result = try await calendar.requestWriteAccessToEvents()
                call.resolve(result)
            }
        }
    }

    @objc public func requestFullCalendarAccess(_ call: CAPPluginCall) {
        Task {
            await tryAsyncCall(call, source: #function) {
                let result = try await calendar.requestFullAccessToEvents()
                call.resolve(["result": result])
            }
        }
    }

    @objc public func requestFullRemindersAccess(_ call: CAPPluginCall) {
        Task {
            await tryAsyncCall(call, source: #function) {
                let result = try await reminders.requestFullAccessToReminders()
                call.resolve(["result": result])
            }
        }
    }

    // Deprecated, use methods above instead.
    @objc public func requestPermission(_ call: CAPPluginCall) {
        guard let alias = getNonEmptyString(call, "alias", source: #function) else {
            return
        }

        Task {
            await tryAsyncCall(call, source: #function) {
                switch alias {
                case "writeCalendar":
                    let result = try await calendar.requestWriteAccessToEvents()
                    call.resolve(result)

                case "readCalendar":
                    let result = try await calendar.requestFullAccessToEvents()
                    call.resolve(["result": result])

                case "writeReminders":
                    let result = try await reminders.requestFullAccessToReminders()
                    call.resolve(["result": result])

                case "readReminders":
                    let result = try await reminders.requestFullAccessToReminders()
                    call.resolve(["result": result])

                default:
                    CapacitorCalendarError.reject(call, type: .invalidKey, source: #function, data: alias)
                    return
                }
            }
        }
    }

    @objc public func requestAllPermissions(_ call: CAPPluginCall) {
        Task {
            await tryAsyncCall(call, source: #function) {
                let calendarResult = try await calendar.requestFullAccessToEvents()
                let remindersResult = try await reminders.requestFullAccessToReminders()
                var result: [String: String] = [
                    "readCalendar": PermissionState.denied.rawValue,
                    "writeCalendar": PermissionState.denied.rawValue,
                    "readReminders": PermissionState.denied.rawValue,
                    "writeReminders": PermissionState.denied.rawValue
                ]

                if calendarResult == PermissionState.granted.rawValue {
                    result["readCalendar"] = PermissionState.granted.rawValue
                    result["writeCalendar"] = PermissionState.granted.rawValue
                }

                if remindersResult == PermissionState.granted.rawValue {
                    result["readReminders"] = PermissionState.granted.rawValue
                    result["writeReminders"] = PermissionState.granted.rawValue
                }

                call.resolve(result)
            }
        }
    }

    // MARK: - Calendars

    @objc public func selectCalendarsWithPrompt(_ call: CAPPluginCall) {
        guard let selectionStyle = call.getInt("selectionStyle") else {
            CapacitorCalendarError.reject(call, type: .missingKey, source: #function, data: "selectionStyle")
            return
        }

        guard let displayStyle = call.getInt("displayStyle") else {
            CapacitorCalendarError.reject(call, type: .missingKey, source: #function, data: "displayStyle")
            return
        }

        Task {
            await tryAsyncCall(call, source: #function) {
                let result = try await calendar.selectCalendarsWithPrompt(selectionStyle: selectionStyle, displayStyle: displayStyle)
                call.resolve(["result": result])
            }
        }
    }

    @objc public func listCalendars(_ call: CAPPluginCall) {
        tryCall(call, source: #function) {
            if try callHasAccess(call, entityType: .calendar, accessType: .read, source: #function) {
                let access = call.getInt("access", EKCalendarChooserDisplayStyle.allCalendars.rawValue) == EKCalendarChooserDisplayStyle.allCalendars.rawValue ? EKCalendarChooserDisplayStyle.allCalendars : EKCalendarChooserDisplayStyle.writableCalendarsOnly
                call.resolve(["result": calendar.listCalendars(access)])
            }
        }
    }

    @objc public func getDefaultCalendar(_ call: CAPPluginCall) {
        tryCall(call, source: #function) {
            if try callHasAccess(call, entityType: .calendar, accessType: .read, source: #function) {
                let defaultCalendar = calendar.getDefaultCalendar()
                call.resolve(["result": defaultCalendar ?? NSNull()])
            }
        }
    }

    @objc public func createCalendar(_ call: CAPPluginCall) {
        guard let title = getNonEmptyString(call, "title", source: #function) else {
            return
        }

        let color = call.getString("color")

        Task {
            await tryAsyncCall(call, source: #function) {
                if try callHasAccess(call, entityType: .calendar, accessType: .write, source: #function) {
                    let id = try await calendar.createCalendar(title: title, color: color)
                    call.resolve(["result": id])
                }
            }
        }
    }

    @objc public func deleteCalendar(_ call: CAPPluginCall) {
        guard let id = getNonEmptyString(call, "id", source: #function) else {
            return
        }

        Task {
            await tryAsyncCall(call, source: #function) {
                if try callHasAccess(call, entityType: .calendar, accessType: .write, source: #function) {
                    try await calendar.deleteCalendar(id: id)
                    call.resolve()
                }
            }
        }
    }

    @objc public func openCalendar(_ call: CAPPluginCall) {
        let interval: Double

        if let date = call.getDouble("date") {
            interval = Date(timeIntervalSince1970: date / 1000).timeIntervalSinceReferenceDate
        } else {
            interval = Date.timeIntervalSinceReferenceDate
        }

        Task {
            await tryAsyncCall(call, source: #function) {
                try await calendar.openCalendar(date: interval)
            }
        }
    }

    // MARK: - Events

    private func doCreateEvent(_ call: CAPPluginCall, prompt: Bool, source: String) {
        Task {
            await tryAsyncCall(call, source: source) {
                if try !callHasAccess(call, entityType: .calendar, accessType: .write, source: source) {
                    return
                }

                let title = call.getString("title", "New event")
                let location = call.getString("location")
                let startDate = call.getDouble("startDate")
                let endDate = call.getDouble("endDate")
                let isAllDay = call.getBool("isAllDay")
                let calendarId = call.getString("calendarId")
                let alertOffsetInMinutes = call.getDouble("alertOffsetInMinutes")

                let eventParameters = EventCreationParameters(
                    title: title,
                    calendarId: calendarId,
                    location: location,
                    startDate: startDate,
                    endDate: endDate,
                    isAllDay: isAllDay,
                    alertOffsetInMinutes: alertOffsetInMinutes
                )

                if prompt {
                    let result = try await calendar.createEventWithPrompt(with: eventParameters)
                    call.resolve(["result": result])
                } else {
                    let result = try await calendar.createEvent(with: eventParameters)
                    call.resolve(["result": result])
                }
            }
        }

    }

    @objc public func createEvent(_ call: CAPPluginCall) {
        if getNonEmptyString(call, "title", source: #function) == nil {
            return
        }

        doCreateEvent(call, prompt: false, source: #function)
    }

    @objc public func createEventWithPrompt(_ call: CAPPluginCall) {
        doCreateEvent(call, prompt: true, source: #function)
    }

    @objc public func listEventsInRange(_ call: CAPPluginCall) {
        guard let startDate = call.getDouble("startDate") else {
            CapacitorCalendarError.reject(call, type: .missingKey, source: #function, data: "startDate")
            return
        }

        guard let endDate = call.getDouble("endDate") else {
            CapacitorCalendarError.reject(call, type: .missingKey, source: #function, data: "endDate")
            return
        }

        tryCall(call, source: #function) {
            if try callHasAccess(call, entityType: .calendar, accessType: .read, source: #function) {
                try call.resolve(["result": calendar.listEventsInRange(startDate: startDate, endDate: endDate)])
            }
        }
    }

    @objc public func deleteEventsById(_ call: CAPPluginCall) {
        guard let eventIds = call.getArray("ids") else {
            CapacitorCalendarError.reject(call, type: .missingKey, source: #function, data: "ids")
            return
        }

        Task {
            await tryAsyncCall(call, source: #function) {
                if try callHasAccess(call, entityType: .calendar, accessType: .write, source: #function) {
                    let deleteResult = try await calendar.deleteEventsById(ids: eventIds)
                    call.resolve([
                        "result": [
                            "deleted": deleteResult.deleted,
                            "failed": deleteResult.failed
                        ]
                    ])
                }
            }
        }
    }

    // MARK: - Reminders

    @objc public func getDefaultRemindersList(_ call: CAPPluginCall) {
        tryCall(call, source: #function) {
            if try callHasAccess(call, entityType: .reminders, accessType: .read, source: #function) {
                let result = reminders.getDefaultRemindersList()
                call.resolve(["result": result ?? NSNull()])
            }
        }
    }

    @objc public func getRemindersLists(_ call: CAPPluginCall) {
        tryCall(call, source: #function) {
            if try callHasAccess(call, entityType: .reminders, accessType: .read, source: #function) {
                call.resolve(["result": reminders.getRemindersLists()])
            }
        }
    }

    @objc public func createReminder (_ call: CAPPluginCall) {
        guard let title = getNonEmptyString(call, "title", source: #function) else {
            return
        }

        let listId = call.getString("listId")
        let priority = call.getInt("priority")
        let isCompleted = call.getBool("isCompleted")
        let startDate = call.getDouble("startDate")
        let dueDate = call.getDouble("dueDate")
        let completionDate = call.getDouble("completionDate")
        let notes = call.getString("notes")
        let url = call.getString("url")
        let location = call.getString("location")
        var recurrence: RecurrenceParameters?

        if let recurrenceData = call.getObject("recurrence") {
            guard let frequency = recurrenceData["frequency"] as? Int else {
                CapacitorCalendarError.reject(call, type: .missingKey, source: #function, data: "frequency, must be provided when using recurrence")
                return
            }

            guard let interval = recurrenceData["interval"] as? Int, interval > 0 else {
                CapacitorCalendarError.reject(call, type: .invalidKey, source: #function, data: "interval, must be greater than 0 when using recurrence")
                return
            }

            let end = recurrenceData["end"] as? Double
            recurrence = RecurrenceParameters(frequency: frequency, interval: interval, end: end)
        }

        let reminderParams = ReminderCreationParameters(
            title: title,
            listId: listId,
            priority: priority,
            isCompleted: isCompleted,
            startDate: startDate,
            dueDate: dueDate,
            completionDate: completionDate,
            notes: notes,
            url: url,
            location: location,
            recurrence: recurrence
        )

        Task {
            await tryAsyncCall(call, source: #function) {
                if try callHasAccess(call, entityType: .reminders, accessType: .write, source: #function) {
                    let id = try await reminders.createReminder(with: reminderParams)
                    call.resolve(["result": id])
                }
            }
        }
    }

    @objc public func openReminders(_ call: CAPPluginCall) {
        Task {
            await tryAsyncCall(call, source: #function) {
                try await reminders.openReminders()
            }
        }
    }
}

File: android/src/main/java/dev/barooni/capacitor/calendar/CapacitorCalendarPlugin.kt
package dev.barooni.capacitor.calendar

import android.Manifest
import android.content.Intent
import android.provider.CalendarContract
import androidx.activity.result.ActivityResult
import com.getcapacitor.JSArray
import com.getcapacitor.JSObject
import com.getcapacitor.Plugin
import com.getcapacitor.PluginCall
import com.getcapacitor.PluginMethod
import com.getcapacitor.annotation.ActivityCallback
import com.getcapacitor.annotation.CapacitorPlugin
import com.getcapacitor.annotation.Permission
import com.getcapacitor.annotation.PermissionCallback
import org.json.JSONObject

@CapacitorPlugin(
    name = "CapacitorCalendar",
    permissions = [
        Permission(
            alias = "readCalendar",
            strings = [
                Manifest.permission.READ_CALENDAR,
            ],
        ),
        Permission(
            alias = "writeCalendar",
            strings = [
                Manifest.permission.WRITE_CALENDAR,
            ],
        ),
    ],
)
class CapacitorCalendarPlugin : Plugin() {
    private var implementation = CapacitorCalendar()
    private var currentSource = ""

    // Utils

    private fun tryCall(
        call: PluginCall,
        source: String,
        block: () -> Unit,
    ) {
        try {
            block()
        } catch (error: Exception) {
            PluginException.reject(call, error, source)
        }
    }

    private fun getNonEmptyString(
        call: PluginCall,
        param: String,
        source: String,
    ): String? {
        val value = call.getString(param)

        if (value.isNullOrEmpty()) {
            PluginException.reject(
                call,
                PluginException.ErrorType.MISSING_KEY,
                source,
                param,
            )
            return null
        } else {
            return value
        }
    }

    // Permissions

    @Throws(PluginException::class)
    private fun callHasAccess(
        call: PluginCall,
        accessType: AccessType,
        source: String,
    ): Boolean {
        val permission =
            doPermissionCheckForAlias("${accessType.name.lowercase()}Calendar", source)

        if (permission == "granted") {
            return true
        } else {
            PluginException.rejectWithNoAccess(
                call,
                EntityType.CALENDAR,
                accessType,
                source,
            )
            return false
        }
    }

    @Throws(PluginException::class)
    private fun doPermissionCheckForAlias(
        alias: String,
        source: String,
    ): String {
        return when (alias) {
            "readCalendar" -> getPermissionState(alias).toString()

            "writeCalendar" -> getPermissionState(alias).toString()

            else -> {
                throw PluginException(
                    PluginException.ErrorType.INTERNAL_ERROR,
                    source,
                    alias,
                )
            }
        }
    }

    @PluginMethod(returnType = PluginMethod.RETURN_PROMISE)
    fun checkPermission(call: PluginCall) {
        val source = ::checkPermission.name

        tryCall(call, source) {
            val permissionName =
                getNonEmptyString(call, "alias", source)
                    ?: return@tryCall

            if (permissionName == "readCalendar" || permissionName == "writeCalendar") {
                call.resolve(JSObject().put("result", getPermissionState(permissionName)))
            } else {
                PluginException.reject(
                    call,
                    PluginException.ErrorType.INVALID_KEY,
                    source,
                    permissionName,
                )
            }
        }
    }

    @PluginMethod(returnType = PluginMethod.RETURN_PROMISE)
    fun checkAllPermissions(call: PluginCall) {
        tryCall(call, ::checkAllPermissions.name) {
            return@tryCall checkPermissions(call)
        }
    }

    private fun doRequestPermission(
        call: PluginCall,
        alias: String,
        source: String,
    ) {
        tryCall(call, source) {
            currentSource = source
            call.data.put("alias", alias)
            return@tryCall requestPermissionForAlias(
                alias,
                call,
                "requestPermissionCallback",
            )
        }
    }

    @Throws(PluginException::class)
    @PermissionCallback
    private fun requestPermissionCallback(call: PluginCall?) {
        if (call == null) {
            throw PluginException(
                PluginException.ErrorType.INTERNAL_ERROR,
                "CapacitorCalendar.$currentSource",
                "Call is not defined",
            )
        }

        val permissionName = call.getString("alias")

        tryCall(call, currentSource) {
            call.resolve(JSObject().put("result", getPermissionState(permissionName)))
        }
    }

    @Deprecated("Use specific request<type>Permission instead")
    @PluginMethod(returnType = PluginMethod.RETURN_PROMISE)
    fun requestPermission(call: PluginCall) {
        val source = ::requestPermission.name
        val alias =
            getNonEmptyString(call, "alias", source)
                ?: return

        doRequestPermission(call, alias, source)
    }

    @PluginMethod(returnType = PluginMethod.RETURN_PROMISE)
    fun requestAllPermissions(call: PluginCall) {
        tryCall(call, "CapacitorCalendar.requestAllPermissions") {
            super.requestPermissions(call)
        }
    }

    @PluginMethod(returnType = PluginMethod.RETURN_PROMISE)
    fun requestReadOnlyCalendarAccess(call: PluginCall) {
        doRequestPermission(
            call,
            "readCalendar",
            ::requestReadOnlyCalendarAccess.name,
        )
    }

    @PluginMethod(returnType = PluginMethod.RETURN_PROMISE)
    fun requestWriteOnlyCalendarAccess(call: PluginCall) {
        doRequestPermission(
            call,
            "writeCalendar",
            ::requestReadOnlyCalendarAccess.name,
        )
    }

    @PluginMethod(returnType = PluginMethod.RETURN_PROMISE)
    fun requestFullCalendarAccess(call: PluginCall) {
        requestAllPermissions(call)
    }

    @PluginMethod(returnType = PluginMethod.RETURN_PROMISE)
    fun requestFullRemindersAccess(call: PluginCall) {
        PluginException.unimplemented(call, ::requestFullRemindersAccess.name)
    }

    // Calendars

    @PluginMethod(returnType = PluginMethod.RETURN_PROMISE)
    fun selectCalendarsWithPrompt(call: PluginCall) {
        PluginException.unimplemented(call, ::selectCalendarsWithPrompt.name)
    }

    @PluginMethod(returnType = PluginMethod.RETURN_PROMISE)
    fun listCalendars(call: PluginCall) {
        val source = ::listCalendars.name

        tryCall(call, source) {
            if (!callHasAccess(call, AccessType.READ, source)) {
                return@tryCall
            }

            val rawAccess = call.getInt("access") ?: CapacitorCalendar.AccessMode.ALL.ordinal
            val access = CapacitorCalendar.AccessMode.fromInt(rawAccess)
            val calendars = implementation.listCalendars(context, access)
            call.resolve(JSObject().put("result", calendars))
        }
    }

    @PluginMethod(returnType = PluginMethod.RETURN_PROMISE)
    fun getDefaultCalendar(call: PluginCall) {
        val source = ::getDefaultCalendar.name

        tryCall(call, source) {
            if (callHasAccess(
                    call,
                    AccessType.READ,
                    source,
                )
            ) {
                val primaryCalendar = implementation.getDefaultCalendar(context)
                call.resolve(JSObject().put("result", primaryCalendar ?: JSONObject.NULL))
            }
        }
    }

    @PluginMethod(returnType = PluginMethod.RETURN_PROMISE)
    fun createCalendar(call: PluginCall) {
        PluginException.unimplemented(call, ::createCalendar.name)
    }

    @PluginMethod(returnType = PluginMethod.RETURN_PROMISE)
    fun deleteCalendar(call: PluginCall) {
        PluginException.unimplemented(call, ::deleteCalendar.name)
    }

    @PluginMethod(returnType = PluginMethod.RETURN_NONE)
    fun openCalendar(call: PluginCall) {
        val timestamp = call.getLong("date") ?: System.currentTimeMillis()

        try {
            return activity.startActivity(implementation.openCalendar(timestamp))
        } catch (error: Exception) {
            call.reject("", "[CapacitorCalendar.${::openCalendar.name}] Unable to open calendar")
        }
    }

    // Events

    @PluginMethod(returnType = PluginMethod.RETURN_PROMISE)
    fun createEventWithPrompt(call: PluginCall) {
        val source = ::createEventWithPrompt.name

        tryCall(call, source) {
            if (!callHasAccess(
                    call,
                    AccessType.WRITE,
                    source,
                )
            ) {
                return@tryCall
            }

            implementation.fetchCalendarEventIDs(context)

            val title = call.getString("title", "")
            val calendarId = call.getString("calendarId")
            val location = call.getString("location")
            val startDate = call.getLong("startDate")
            val endDate = call.getLong("endDate")
            val isAllDay = call.getBoolean("isAllDay", false)

            val intent = Intent(Intent.ACTION_INSERT).setData(CalendarContract.Events.CONTENT_URI)

            intent.putExtra(CalendarContract.Events.TITLE, title)
            calendarId?.let { intent.putExtra(CalendarContract.Events.CALENDAR_ID, it) }
            location?.let { intent.putExtra(CalendarContract.Events.EVENT_LOCATION, it) }
            startDate?.let { intent.putExtra(CalendarContract.EXTRA_EVENT_BEGIN_TIME, it) }
            endDate?.let { intent.putExtra(CalendarContract.EXTRA_EVENT_END_TIME, it) }
            isAllDay?.let {
                intent.putExtra(
                    CalendarContract.EXTRA_EVENT_ALL_DAY,
                    if (it) 1 else 0,
                )
            }

            return@tryCall startActivityForResult(
                call,
                intent,
                "openCalendarIntentActivityCallback",
            )
        }
    }

    @Throws(PluginException::class)
    @ActivityCallback
    private fun openCalendarIntentActivityCallback(
        call: PluginCall?,
        result: ActivityResult,
    ) {
        if (call == null) {
            throw PluginException(
                PluginException.ErrorType.INTERNAL_ERROR,
                "CapacitorCalendar.${::createEventWithPrompt.name}",
                "call is not defined",
            )
        }

        val newEventIds =
            implementation.getNewEventIds(implementation.fetchCalendarEventIDs(context))
        val newIdsArray = JSArray()
        newEventIds.forEach { id -> newIdsArray.put(id.toString()) }
        call.resolve(JSObject().put("result", newIdsArray))
    }

    @PluginMethod(returnType = PluginMethod.RETURN_PROMISE)
    fun createEvent(call: PluginCall) {
        val source = ::createEvent.name

        tryCall(call, source) {
            if (!callHasAccess(call, AccessType.WRITE, source)) {
                return@tryCall
            }

            val title =
                getNonEmptyString(call, "title", source)
                    ?: return@tryCall
            val calendarId = call.getString("calendarId")
            val location = call.getString("location")
            val startDate = call.getLong("startDate")
            val endDate = call.getLong("endDate")
            val isAllDay = call.getBoolean("isAllDay", false)
            val alertOffsetInMinutes = call.getFloat("alertOffsetInMinutes")

            val eventUri =
                implementation.createEvent(
                    context,
                    title,
                    calendarId,
                    location,
                    startDate,
                    endDate,
                    isAllDay,
                    alertOffsetInMinutes,
                )

            val id =
                eventUri?.lastPathSegment
                    ?: throw PluginException(
                        PluginException.ErrorType.OS_ERROR,
                        source,
                        "Failed to insert event into calendar",
                    )
            call.resolve(JSObject().put("result", id))
        }
    }

    @PluginMethod(returnType = PluginMethod.RETURN_PROMISE)
    fun listEventsInRange(call: PluginCall) {
        val source = ::listEventsInRange.name

        tryCall(call, source) {
            if (!callHasAccess(
                    call,
                    AccessType.READ,
                    source,
                )
            ) {
                return@tryCall
            }

            val startDate =
                call.getLong("startDate")
                    ?: throw PluginException(
                        PluginException.ErrorType.MISSING_KEY,
                        source,
                        "startDate",
                    )
            val endDate =
                call.getLong("endDate")
                    ?: throw PluginException(
                        PluginException.ErrorType.MISSING_KEY,
                        source,
                        "endDate",
                    )
            val ret = JSObject()
            ret.put("result", implementation.listEventsInRange(context, startDate, endDate))
            call.resolve(ret)
        }
    }

    @PluginMethod(returnType = PluginMethod.RETURN_PROMISE)
    fun deleteEventsById(call: PluginCall) {
        val source = ::deleteEventsById.name

        tryCall(call, source) {
            if (!callHasAccess(
                    call,
                    AccessType.WRITE,
                    source,
                )
            ) {
                return@tryCall
            }

            val ids =
                call.getArray("ids")
                    ?: throw PluginException(
                        PluginException.ErrorType.MISSING_KEY,
                        source,
                        "ids",
                    )
            call.resolve(JSObject().put("result", implementation.deleteEventsById(context, ids)))
        }
    }

    // Reminders

    @PluginMethod(returnType = PluginMethod.RETURN_PROMISE)
    fun getDefaultRemindersList(call: PluginCall) {
        PluginException.unimplemented(call, ::getDefaultRemindersList.name)
    }

    @PluginMethod(returnType = PluginMethod.RETURN_PROMISE)
    fun getRemindersLists(call: PluginCall) {
        PluginException.unimplemented(call, ::getRemindersLists.name)
    }

    @PluginMethod(returnType = PluginMethod.RETURN_PROMISE)
    fun createReminder(call: PluginCall) {
        PluginException.unimplemented(call, ::createReminder.name)
    }

    @PluginMethod(returnType = PluginMethod.RETURN_PROMISE)
    fun openReminders(call: PluginCall) {
        PluginException.unimplemented(call, ::openReminders.name)
    }
}
